src/lib/Microsoft.Health.Events/EventConsumers/Service/EventBatchingService.cs

```diff
@@ -73,6 +80,75 @@ public void NewPartitionInitialized(string partitionId)
             {
                 _eventPartitions.TryRemove(partitionId, out _);
             }
+
+            IncrementPartitionCount();
+
+            // if _scaledOutMaxEventsAllPartitions is set then assume we want to adjust the batch size based on
+            // how many partitions are currently held by the processor.
+            if (_scaledOutMaxEventsAllPartitions != null)
+            {
+                ChangeMaximumEventsPerPartition();
+            }
+        }
+
+        public void PartitionProcessingStopped(string partitionId)
+        {
+            if (EventPartitionExists(partitionId))
+            {
+                _eventPartitions.TryRemove(partitionId, out _);
+            }
+
+            DecrementPartitionCount();
+
+            if (_scaledOutMaxEventsAllPartitions != null)
+            {
+                ChangeMaximumEventsPerPartition();
+            }
+        }
+
+        private void ChangeMaximumEventsPerPartition()
+        {
+            // If 1 partition is held then allocate all events to that partition.
+            // If 2 partitions are held then allocate half of the events to one partition and half to the other partition
+            // etc
+            // If the result would be lower than MaxEvents, then use MaxEvents.
+            // Use locking to prevent multiple threads from changing the maximum events value at the same time.
+            lock (_lock)
+            {
+                _logger.LogTrace($"Currently processing {_partitionCount} partitions");
+                var newMaxEventsValue = Math.Max(_scaledOutMaxEventsAllPartitions.Value / Math.Max(_partitionCount, 1), _maxEvents);
+                var oldMaxEventsValue = _scaledOutMaxEventsPerPartition;
+                _logger.LogTrace($"Updating maximum events from {oldMaxEventsValue} to {newMaxEventsValue}");
+                _scaledOutMaxEventsPerPartition = newMaxEventsValue;
+            }
+        }
+
+        private void IncrementPartitionCount()
+        {
+            lock (_lock)
+            {
+                _partitionCount++;
+            }
+        }
+
+        private void DecrementPartitionCount()
+        {
+            lock (_lock)
+            {
+                _partitionCount--;
+            }
+        }
+
+        public int GetMaximumEventsForPartition()
+        {
+            if (_scaledOutMaxEventsPerPartition <= 0)
```

Comment:
This should also have some thread safety for access as well.

Reply:
We could add some thread safety but I chose not too due to the potential performance impact of locking/interlocking to read the variable and perform the check on each event to determine if we should flush or not. It should be fine to read the variable and perform the flush (or not perform the flush) based on whatever the value this method returns, even if the method returned a value that was slightly out of date.

The one condition we did need to account for is if the _scaledOutMaxEventsPerPartition were to increase such that _scaledOutMaxEventsPerPartition is greater than the current amount of events in the queue for the partition but in this PR have updated the code to account for that.

Reply:
That is fair. It would be good to add comment on the property though explaining why we aren't applying thread safety so it doesn't appear like a bug when someone is looking at this later.